src/Microsoft.Data.SqlClient/add-ons/AzureKeyVaultProvider/AzureKeyVaultProviderTokenCredential.cs

```diff
@@ -0,0 +1,90 @@
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+
+namespace Microsoft.Data.SqlClient.AlwaysEncrypted.AzureKeyVaultProvider
+{
+    internal class AzureKeyVaultProviderTokenCredential : TokenCredential
+    {
+        private const string Bearer = "Bearer ";
+
+        private AuthenticationCallback Callback { get; set; }
+
+        private string Authority { get; set; }
+
+        private string Resource { get; set; }
+
+        internal AzureKeyVaultProviderTokenCredential(AuthenticationCallback authenticationCallback, string masterKeyPath)
+        {
+            Callback = authenticationCallback;
+            using (HttpClient httpClient = new HttpClient())
+            {
+                HttpResponseMessage response = httpClient.GetAsync(masterKeyPath).GetAwaiter().GetResult();
+                string challenge = response?.Headers.WwwAuthenticate.FirstOrDefault()?.ToString();
+                string trimmedChallenge = ValidateChallenge(challenge);
+                string[] pairs = trimmedChallenge.Split(new string[] { "," }, StringSplitOptions.RemoveEmptyEntries);
+
+                if (pairs != null && pairs.Length > 0)
+                {
+                    for (int i = 0; i < pairs.Length; i++)
+                    {
+                        string[] pair = pairs[i]?.Split('=');
+
+                        if (pair.Length == 2)
+                        {
+                            string key = pair[0]?.Trim().Trim(new char[] { '\"' });
+                            string value = pair[1]?.Trim().Trim(new char[] { '\"' });
+
+                            if (!string.IsNullOrEmpty(key))
+                            {
+                                if (key.Equals("authorization", StringComparison.InvariantCultureIgnoreCase))
+                                {
+                                    Authority = value;
+                                }
+                                else if (key.Equals("resource", StringComparison.InvariantCultureIgnoreCase))
+                                {
+                                    Resource = value;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        public override AccessToken GetToken(TokenRequestContext requestContext, CancellationToken cancellationToken)
+        {
+            string token = Callback.Invoke(Authority, Resource, string.Empty).GetAwaiter().GetResult();
+            return new AccessToken(token, DateTimeOffset.Now.AddHours(1));
+        }
+
+        public override ValueTask<AccessToken> GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken)
+        {
+            string token = Callback.Invoke(Authority, Resource, string.Empty).GetAwaiter().GetResult();
```

Comment:
Why not just declare an `async` method and `await` the results here? In async library code, you most often should call `ConfigureAwait(false)` to avoid potential hangs in calling applications.

Reply:
This is the callback received from `SqlColumnEncryptionAzureKeyVaultProvider` [constructor](https://github.com/dotnet/SqlClient/pull/630/files#diff-b402b979fad98d38b35b0522ab2fabb6R105) - to support old KeyVaultClient.AuthenticationCallback, that we invoke here to acquire accesstoken (string).